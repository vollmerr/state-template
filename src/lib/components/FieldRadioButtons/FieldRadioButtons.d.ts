import * as React from 'react';

import * as types from '../../utils/types';

export interface FieldRadioButtonsProps {
  /** Accessible indicator of related information */
  'aria-describedby'?: string;

  /** Accessible indicator for errors existing */
  'aria-invalid'?: string;

  /** Class names to attach to the field */
  className?: string;

  /** Disable the input */
  disabled?: boolean;

  /** Id of field */
  id?: string;

  /** Display inline */
  inline?: boolean;

  /** Label for the field */
  label: string;

  /** Name of field */
  name: string;

  /** Called when the field is blurred */
  onBlur?: (event: FocusEvent, value?: string) => void;

  /** Called when the field changes */
  onChange?: (event: Event, value?: string) => void;

  /** Called when the field is focused */
  onFocus?: (event: FocusEvent, nodeType: string) => void;

  /** Options to select from */
  options: Array<types.Option>;

  /** Determines if field is required */
  required?: boolean;

  /** Tooltip to render */
  tooltip?: React.ReactNode;

  /** Value of the field */
  value?: string;

  /** Use style variant */
  variant?: 'highlight';
}

declare class FieldRadioButtons extends React.Component<FieldRadioButtonsProps, {}> {}

export default FieldRadioButtons;
