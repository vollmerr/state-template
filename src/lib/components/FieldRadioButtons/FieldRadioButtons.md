```jsx
const initialValues = {
  radioButton: 'value2',
  disabledRadioButton: 'value2',
};

<ExampleForm form={'radioButton'} initialValues={initialValues}>
  <FieldRadioButtons 
    name={'radioButton'}
    label={'Radio Button'}
    options={[
      { value: 'value1', label: 'label 1' },
      { value: 'value2', label: 'label 2' },
    ]}
    helpText={'This is the `helpText` content...'}
  />
  <FieldRadioButtons 
    required
    name={'requiredRadioButton'}
    label={'Required Radio Button'}
    options={[
      { value: 'value1', label: 'label 1' },
      { value: 'value2', label: 'label 2' },
    ]}
  />
  <FieldRadioButtons 
    disabled
    name={'disabledRadioButton'}
    label={'Disabled Radio Button'}
    options={[
      { value: 'value1', label: 'label 1' },
      { value: 'value2', label: 'label 2' },
    ]}
  /> 
</ExampleForm>
```

### inline

```jsx
const initialValues = {
  inlineRadioButtons: 'value2',
};

<ExampleForm form={'inlineRadioButton'} initialValues={initialValues}>
  <FieldRadioButtons 
    inline
    name={'inlineRadioButtons'}
    label={'Inline Radio Buttons'}
    options={[
      { value: 'value1', label: 'label 1' },
      { value: 'value2', label: 'label 2' },
      { value: 'value3', label: 'label 3' },
    ]}
  />
</ExampleForm>
```

### event handlers

```jsx
const initialValues = {
  handlersRadioButtons: 'value2',
};

<ExampleForm form={'handlersRadioButton'} initialValues={initialValues}>
  <FieldRadioButtons 
    name={'handlersRadioButtons'}
    label={'With Event Handlers'}
    options={[
      { value: 'value1', label: 'label 1' },
      { value: 'value2', label: 'label 2' },
    ]}
    onBlur={(value) => { console.log('blurring: ', value); }}
    onChange={(value) => { console.log('changing: ', value); }}
    onFocus={(value) => { console.log('focusing: ', value); }}
  />
</ExampleForm>
```

### highlight variant

```jsx
const initialValues = {
  primaryRadioButton: 'value2',
  requiredPrimaryRadioButton: 'value2',
  disabledPrimaryRadioButton: 'value2',
};

<ExampleForm form={'radioButtonPrimary'} initialValues={initialValues}>
  <FieldRadioButtons 
    variant={'highlight'}
    name={'primaryRadioButton'}
    label={'Primary Radio Button'}
    options={[
      { value: 'value1', label: 'label 1' },
      { value: 'value2', label: 'label 2' },
    ]}
    helpText={'This is the `helpText` content...'}
  />
  <FieldRadioButtons 
    required
    variant={'highlight'}
    name={'requiredPrimaryRadioButton'}
    label={'Required Primary Radio Button'}
    options={[
      { value: 'value1', label: 'label 1' },
      { value: 'value2', label: 'label 2' },
    ]}
  />
  <FieldRadioButtons 
    disabled
    variant={'highlight'}
    name={'disabledPrimaryRadioButton'}
    label={'Disabled Primary Radio Button'}
    options={[
      { value: 'value1', label: 'label 1' },
      { value: 'value2', label: 'label 2' },
    ]}
  />
</ExampleForm>
```
