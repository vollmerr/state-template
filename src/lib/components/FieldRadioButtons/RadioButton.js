import React from 'react';
import T from 'prop-types';
import classNames from 'classnames';

import * as propUtils from '../../utils/propTypes';

// individual radio button, applies state-template styling
const RadioButton = (props) => {
  const {
    name,
    value,
    option,
    inline,
    variant,
    disabled,
    onChange,
    className,
    'aria-invalid': ariaInvalid,
    'aria-describedby': ariaDescribedby,
  } = props;

  const id = `${name}-${option.value}`;
  const inputProps = {
    id,
    name,
    disabled,
    type: 'radio',
    checked: value === option.value,
    onChange: () => onChange(option.value),
    'aria-invalid': ariaInvalid,
    'aria-describedby': ariaDescribedby,
  };

  let cn = classNames([
    'check',
    className,
  ]);

  if (variant) {
    return (
      <label data-test={'field--radio-button'} className={cn} htmlFor={id}>
        <input className={'hidden-up pos-abs top-0 left-0'} {...inputProps} />
        <span className={`btn btn-lg btn-block bg-${variant}--checked rounded-0`}>
          <div>{option.label}</div>
        </span>
      </label>
    );
  }

  cn = classNames([
    'p-l-md m-l-0 m-r-md',
    className,
    inline ? 'form-check-inline' : 'form-check',
  ]);

  return (
    <label data-test={'field--radio-button'} className={cn} htmlFor={id}>
      <input id={id} className={'hidden-up pos-abs'} {...inputProps} />
      <div className={'check-icon-radio'} aria-hidden>
        <i />
      </div>
      <div>{option.label}</div>
    </label>
  );
};

RadioButton.propTypes = {
  /** Name of field */
  name: T.string.isRequired,

  /** Value of radio button */
  value: T.string.isRequired,

  /** Option to select */
  option: propUtils.option.isRequired,

  /** Display inline */
  inline: T.bool,

  /** Use style variant */
  variant: T.oneOf([
    '',
    'highlight',
  ]),

  /** Disable the input */
  disabled: T.bool,

  /** Called when radio button changes */
  onChange: T.func.isRequired,

  /** Class name to attach to top level label */
  className: T.string,

  /** Accessible indicator for errors existing */
  'aria-invalid': T.string.isRequired,

  /** Accessible indicator of related information */
  'aria-describedby': T.string,
};

RadioButton.defaultProps = {
  inline: false,
  variant: '',
  disabled: false,
  className: '',
  'aria-describedby': null,
};

export default RadioButton;
