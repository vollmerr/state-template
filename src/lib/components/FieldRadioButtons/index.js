export { default } from './FieldRadioButtons';
export { default as RadioButton } from './RadioButton';
