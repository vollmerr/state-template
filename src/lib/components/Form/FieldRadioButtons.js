import React from 'react';

import FieldWrapper from './FieldWrapper';
import withField from './withField';

// individual radio button, applies state-template styling
const RadioButton = ({
  input, value, label, styleType,
}) => {
  const checked = input.value.toString().indexOf(value) !== -1;
  const onChange = () => input.onChange(value);

  if (styleType === 'button') {
    return (
      <label className={'check'}>
        <input className={'hidden-up pos-abs top-0 left-0'} name={input.name} checked={checked} onChange={onChange} type={'radio'} />
        <span className={'btn btn-lg btn-block bg-highlight--checked rounded-0'}>{label}</span>
      </label>
    );
  }

  return (
    <label className={'form-check-inline p-l-md m-l-0 m-r-md'}>
      <input className={'hidden-up pos-abs'} name={input.name} checked={checked} onChange={onChange} type={'radio'} />
      <div className={'check-icon-radio'}>
        <i />
      </div>
      {label}
    </label>
  );
};

// group of radio buttons with optional help text and label
// displays error if validation fails
const FieldRadioButtons = (props) => {
  const { input, options, styleType } = props;

  return (
    <FieldWrapper {...props}>
      <div>
        {
          options.filter(x => !x.hidden).map(option => (
            <RadioButton
              input={input}
              key={option.value}
              styleType={styleType}
              {...option}
            />
          ))
        }
      </div>
    </FieldWrapper>
  );
};

export default withField(FieldRadioButtons);
