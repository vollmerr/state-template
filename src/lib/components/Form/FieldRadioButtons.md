```jsx


<ExampleForm form={'radioButton'}>
  <FieldRadioButtons 
    name={'radioButton'}
    label={'Radio Button'}
    options={[
      { value: 'value 1', label: 'label 1' },
      { value: 'value 2', label: 'label 2' },
    ]}
    helpText={'This is the `helpText` content...'}
  />
  <FieldRadioButtons 
    required
    name={'requiredRadioButton'}
    label={'Required Radio Button'}
    options={[
      { value: 'value 1', label: 'label 1' },
      { value: 'value 2', label: 'label 2' },
    ]}
  />
  <FieldRadioButtons 
    disabled
    name={'disabledRadioButton'}
    label={'Disabled Radio Button'}
    options={[
      { value: 'value 1', label: 'label 1' },
      { value: 'value 2', label: 'label 2' },
    ]}
  />
</ExampleForm>
```

### highlight variant

```jsx
<ExampleForm form={'radioButtonPrimary'}>
  <FieldRadioButtons 
    variant={'highlight'}
    name={'primaryRadioButton'}
    label={'Primary Radio Button'}
    options={[
      { value: 'value 1', label: 'label 1' },
      { value: 'value 2', label: 'label 2' },
    ]}
    helpText={'This is the `helpText` content...'}
  />
  <FieldRadioButtons 
    required
    variant={'highlight'}
    name={'requiredPrimaryRadioButton'}
    label={'Required Primary Radio Button'}
    options={[
      { value: 'value 1', label: 'label 1' },
      { value: 'value 2', label: 'label 2' },
    ]}
  />
  <FieldRadioButtons 
    disabled
    variant={'highlight'}
    name={'disabledPrimaryRadioButton'}
    label={'Disabled Primary Radio Button'}
    options={[
      { value: 'value 1', label: 'label 1' },
      { value: 'value 2', label: 'label 2' },
    ]}
  />
</ExampleForm>
```
