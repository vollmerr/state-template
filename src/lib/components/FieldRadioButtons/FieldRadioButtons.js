import React from 'react';
import T from 'prop-types';
import classNames from 'classnames';

import { isEmptyRadio } from '../../utils/validate';
import * as propUtils from '../../utils/propTypes';
import { withField } from '../Field';

import RadioButton from './RadioButton';

// group of radio buttons
export const FieldRadioButtons = (props) => {
  const {
    'aria-describedby': ariaDescribedby,
    'aria-invalid': ariaInvalid,
    className,
    disabled,
    id,
    inline,
    label,
    name,
    onBlur,
    onChange,
    onFocus,
    options,
    value,
    variant,
    ...rest
  } = props;

  const cn = classNames([
    'field__radio-buttons',
    className,
  ]);

  const wrapperProps = {
    'aria-label': label,
    'data-test': 'field__radio-buttons',
    className: cn,
    id,
    role: 'group',
    ...rest,
  };

  const radioProps = {
    'aria-describedby': ariaDescribedby,
    'aria-invalid': ariaInvalid,
    disabled,
    inline,
    name,
    onBlur,
    onChange,
    onFocus,
    value,
    variant,
  };

  return (
    <div {...wrapperProps}>
      {
        options.filter((x) => !x.hidden).map((option) => (
          <RadioButton
            key={option.value}
            option={option}
            {...option}
            {...radioProps}
          />
        ))
      }
    </div>
  );
};

FieldRadioButtons.propTypes = {
  /** Accessible indicator of related information */
  'aria-describedby': T.string,

  /** Accessible indicator for errors existing */
  'aria-invalid': T.string,

  /** Class names to attach to the field */
  className: T.string,

  /** Disable the input */
  disabled: T.bool,

  /** Id of field */
  id: T.string,

  /** Display inline */
  inline: T.bool,

  /** aria-label for the field */
  label: T.string.isRequired,

  /** Name of field */
  name: T.string.isRequired,

  /** Called when radio button is blurred */
  onBlur: T.func,

  /** Called when radio button changes */
  onChange: T.func,

  /** Called when radio button is focused */
  onFocus: T.func,

  /** Options to select from */
  options: T.arrayOf(propUtils.option).isRequired,

  /** Value of the field */
  value: T.string,

  /** Use style variant */
  variant: T.oneOf([
    'highlight',
  ]),
};

FieldRadioButtons.defaultProps = {
  'aria-describedby': null,
  'aria-invalid': 'false',
  className: null,
  disabled: false,
  id: null,
  inline: false,
  onBlur: null,
  onChange: null,
  onFocus: null,
  value: null,
  variant: null,
};

export default withField(isEmptyRadio)(FieldRadioButtons);
