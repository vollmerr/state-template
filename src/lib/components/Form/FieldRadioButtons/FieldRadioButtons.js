import React from 'react';
import T from 'prop-types';

import { isEmptyRadio } from '../../../utils/validate';

import FieldWrapper from '../FieldWrapper';
import withField from '../withField';
import RadioButton from './RadioButton';

// group of radio buttons with optional help text and label
// displays error if validation fails
const FieldRadioButtons = (props) => {
  const {
    input, options, variant, disabled, ...rest
  } = props;

  return (
    <FieldWrapper data-test={'field-radio-buttons'} disabled={disabled} {...rest}>
      <div className={'field-radio'}>
        {
          options.filter(x => !x.hidden).map(option => (
            <RadioButton
              input={input}
              key={option.value}
              variant={variant}
              disabled={disabled}
              {...option}
            />
          ))
        }
      </div>
    </FieldWrapper>
  );
};


FieldRadioButtons.propTypes = {
  /** Use style variant */
  variant: T.oneOf([
    '',
    'highlight',
  ]),
  /** Input from redux-form's Field, attaches name, value, etc */
  input: T.object.isRequired,
  /** Options to select from */
  options: T.arrayOf(
    T.shape({
      label: T.string.isRequired,
      value: T.string.isRequired,
    }),
  ).isRequired,
  /** Disable the input */
  disabled: T.bool,
};

FieldRadioButtons.defaultProps = {
  variant: '',
  disabled: false,
};

const withReduxField = withField(isEmptyRadio);
export default withReduxField(FieldRadioButtons);
