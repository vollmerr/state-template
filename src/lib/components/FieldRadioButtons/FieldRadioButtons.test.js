import React from 'react';
import { shallow } from 'enzyme';

import RadioButton from './RadioButton';
import { FieldRadioButtons } from './FieldRadioButtons';

const { testUtils } = global;

const props = {
  input: testUtils.getInputProp(),
  meta: testUtils.getMetaProp(),
  label: 'test label',
  options: [
    { value: 'v1', label: 'l1' },
    { value: 'v2', label: 'l2' },
  ],
  disabled: false,
};

let wrapper;
describe('FieldRadioButtons', () => {
  beforeEach(() => {
    wrapper = shallow(<FieldRadioButtons {...props} />);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render all the options as `RadioButton`s', () => {
    expect(wrapper.find(RadioButton).length).toBe(props.options.length);
  });
});
