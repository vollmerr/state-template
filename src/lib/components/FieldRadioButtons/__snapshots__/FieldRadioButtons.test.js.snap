// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldRadioButtons should render correctly 1`] = `
<Fragment>
  <FieldLabel
    id="test-id--label"
    label="test-label"
    name="test-name"
    required={false}
    tag="div"
    tooltip={null}
  />
  <div
    aria-describedby="test-id--label"
    className="field__radio-buttons test-classname"
    data-test="field__radio-buttons"
    id="test-id"
    role="radiogroup"
  >
    <RadioButton
      aria-describedby="aria-desc-id"
      aria-invalid="false"
      className="option-class"
      disabled={false}
      inline={false}
      key="value-1"
      label="label-1"
      name="test-name"
      onBlur={[MockFunction]}
      onChange={[MockFunction]}
      onFocus={[MockFunction]}
      option={
        Object {
          "className": "option-class",
          "label": "label-1",
          "value": "value-1",
        }
      }
      value="value-1"
      variant={null}
    />
    <RadioButton
      aria-describedby="aria-desc-id"
      aria-invalid="false"
      disabled={false}
      inline={false}
      key="value-2"
      label="label-2"
      name="test-name"
      onBlur={[MockFunction]}
      onChange={[MockFunction]}
      onFocus={[MockFunction]}
      option={
        Object {
          "label": "label-2",
          "value": "value-2",
        }
      }
      value="value-1"
      variant={null}
    />
  </div>
</Fragment>
`;
